### ISSUE TITLE
Some users have a broken user document

### ISSUE THREAD

**Issue by nilsreichardt** [SharezoneApp/sharezone-app#48]
![image](https://user-images.githubusercontent.com/24459435/134209171-eb9755b6-c018-4f44-bf7d-387aa77c026f.png)

I don't how, but the user `5VLQnu7xwSNfVjIiyXk6JRef22N2` has a broken user document. Fields like `typeOfUser` are missing. Therefore, this user can not see any homeworks, because the cloud function, which assigns the users to homework, do not have the `typeOfUser`.

**Comment by Jonas-Sander** [comment]
Is there something actionable we can do here really? Do you have information on why it might be missing?

**Comment by nilsreichardt** [comment]
No, I don't have any idea, why this happen.

What we can do is catching the missing typeOfUser behaviour in the cloud function, which adds the user ids and check, if there are users without a typeOfUser. If yes, we notify ourselves via Slack. Advantage of this: Getting a feeling how many users are affected and getting the user ids of the affected users (they could give us more information, what they did before).

**Comment by Jonas-Sander** [comment]
@nilsreichardt Did you investigate this? Or is this still important? Just asking because this is still in "Soon-ish" in the triage board.

**Comment by nilsreichardt** [comment]
I was reproducible with one of the versions before the summer (by using Google or Apple Sign In with an account that is not registered) but was fixed with one the updates in the summer. However, we should still do the following tasks:

* Check if there are still broken documents even after the new updates (making a query that check the user documents of the last weeks)
* Write a script that fixes the broken user documents (just using some default values because with a broken user document some features are not working like renaming)

**Comment by nilsreichardt** [comment]
# TL;DR

This issue appears to have been resolved. The last user with a broken user document created their account on 2023-10-31. Additionally, I have fixed all broken user documents that were in our database.

# Here is the documentation of how I did it:

Because we can't query for documents without a field, I exported the `User` collection using:

```sh
PROJECT_ID="sharezone-c2bd8"
gcloud firestore export gs://$PROJECT_ID.appspot.com/export/2024-02-27-users \
    --collection-ids=User \
    --project=$PROJECT_ID
```

Then I imported the data to BigQuery using the following instructions: https://cloud.google.com/bigquery/docs/loading-data-cloud-firestore#console

I used the following SQL query to get all affected users:

```sql
SELECT __key__.name as UserId, createdOn, typeOfUser
FROM `sharezone-c2bd8.2024_02_27_User.User`
WHERE typeOfUser IS NULL
ORDER BY createdOn DESC
```

The number of affected users is 2251.

I exported the SQL query result as JSON and executed the following script:

```ts
import * as fs from "fs";
import { firestore } from "../firebase_globals";

async function main() {
  // Read users from JSON file
  const data = fs.readFileSync(`${__dirname}/data.json`).toString();
  const json = JSON.parse(data);
  let counter = 0;

  for (const user of json) {
    // Update user in Firestore
    await firestore.doc(`User/${user.UserId}`).update({
      typeOfUser: "student", // assuming that every affacted user is a student
    });

    // Log the action
    const logMessage = `Updated user ${user.UserId} to typeOfUser "student"`;
    fs.appendFileSync("log.txt", `${logMessage}\n`);
    console.log(`${logMessage}; ${++counter} users updated`);
  }
}

main().catch(console.error);
```

It's enough to only set the `typeOfUser`. Some cloud functions require a `typeOfUser`, like the `UserUpdate` cloud function. The other fields (name, etc.) can be easily changed by the user in the settings.
